src/realm/object-store/sync/sync_user.cpp

```diff
@@ -147,79 +167,41 @@ std::shared_ptr<SyncManager> SyncUser::sync_manager() const
             util::format("Cannot start a sync session for user '%1' because this user has been removed.",
                          m_identity));
     }
-    REALM_ASSERT(m_sync_manager);
-    return m_sync_manager->shared_from_this();
+    return m_app;
 }
 
-void SyncUser::detach_from_sync_manager()
+void SyncUser::detach_from_backing_store()
 {
     util::CheckedLockGuard lk(m_mutex);
-    REALM_ASSERT(m_sync_manager);
     m_state = SyncUser::State::Removed;
-    m_sync_manager = nullptr;
+    m_app.reset();
 }
 
-std::vector<std::shared_ptr<SyncSession>> SyncUser::all_sessions()
-{
-    util::CheckedLockGuard lock(m_mutex);
-    std::vector<std::shared_ptr<SyncSession>> sessions;
-    if (m_state == State::Removed) {
-        return sessions;
-    }
-    for (auto it = m_sessions.begin(); it != m_sessions.end();) {
-        if (auto ptr_to_session = it->second.lock()) {
-            sessions.emplace_back(std::move(ptr_to_session));
-            it++;
-            continue;
-        }
-        // This session is bad, destroy it.
-        it = m_sessions.erase(it);
-    }
-    return sessions;
-}
-
-std::shared_ptr<SyncSession> SyncUser::session_for_on_disk_path(const std::string& path)
-{
-    util::CheckedLockGuard lock(m_mutex);
-    if (m_state == State::Removed) {
-        return nullptr;
-    }
-    auto it = m_sessions.find(path);
-    if (it == m_sessions.end()) {
-        return nullptr;
-    }
-    auto locked = it->second.lock();
-    if (!locked) {
-        // Remove the session from the map, because it has fatally errored out or the entry is invalid.
-        m_sessions.erase(it);
-    }
-    return locked;
-}
-
-void SyncUser::log_in(const std::string& access_token, const std::string& refresh_token)
+void SyncUser::log_in(std::string_view access_token, std::string_view refresh_token)
 {
     REALM_ASSERT(!access_token.empty());
     REALM_ASSERT(!refresh_token.empty());
-    std::vector<std::shared_ptr<SyncSession>> sessions_to_revive;
     {
         util::CheckedLockGuard lock1(m_mutex);
         util::CheckedLockGuard lock2(m_tokens_mutex);
         m_state = State::LoggedIn;
         m_access_token = RealmJWT(access_token);
         m_refresh_token = RealmJWT(refresh_token);
-        sessions_to_revive = revive_sessions();
 
-        m_sync_manager->perform_metadata_update([&](const auto& manager) {
+        lock_or_throw(m_app)->backing_store()->perform_metadata_update([&](const auto& manager) {
             auto metadata = manager.get_or_make_user_metadata(m_identity);
             metadata->set_state_and_tokens(State::LoggedIn, access_token, refresh_token);
         });
     }
+#if REALM_ENABLE_SYNC
+    std::vector<std::shared_ptr<SyncSession>> sessions_to_revive = revive_sessions();
```

Comment:
There's either incorrect locking here or remnants of no longer needed locking. I think it's the latter and the locking has been pushed elsewhere so this code isn't incorrect, but it's now more confusing than it needs to be. The comment below is no longer applicable, and revive_sessions() doesn't need to copy the vector into a different vector.

Reply:
I think reviving sessions needs to be done outside the user locks still, but I've simplified this.